src/train.py

```diff
@@ -16,67 +16,248 @@
 
-def eval_text_auto(target, hidden_state, enc_output, enc_ctxt_mask, text_model):
-    res = text_model.infer_sequence(hidden_state, enc_output, enc_ctxt_mask, max_len=target.shape[1])
-    loss = loss_fn(res, target).item()
-    return loss
+def discriminator_loss(output, target):
+    return F.cross_entropy(output, target)
 
-def eval_speech_auto(target, hidden_state, enc_output, enc_ctxt_mask, text_model):
-    res = text_model.infer_sequence(hidden_state, enc_output, enc_ctxt_mask, max_len=target.shape[1])
-    loss = loss_fn(res, target).item()
-    return loss
 
-def evaluate(text_model, speech_model, valid_dataset):
+def evaluate(model, valid_dataset):
     """
         Expect validation set to have paired speech & text!
-        We evaluate on 4 metrics:  autoencoder text loss, autoencoder speech loss,
-            ASR (evaluated by PER), and TTS (evaluated by MSE). 
+        Primary evaluation is PER - can gauge training by the other losses
+        We return on 6 other metrics:  autoencoder text loss, autoencoder speech loss,
+            ASR loss, and TTS loss, cross model text and cross model speech loss. 
     """
-    text_model.eval()
-    speech_model.eval()
+    model.eval()
     with torch.no_grad():
-        ae_text_loss, ae_speech_loss, per, tts_err = 0, 0, 0, 0
-        avg_constant = 0
-        for data in valid_dataset:
-            character, mel, mel_input, pos_text, pos_mel, text_len = data
-            character = character.to(DEVICE)
-            mel = mel.to(DEVICE)
-            enc_text, text_hidden_state, text_pad_mask = text_model.encode(character)
-            enc_speech, speech_hidden_state, speech_pad_mask = speech_model.encode(mel)
+        losses = defaultdict(list)
+        per, n_iters = 0, 0
+        
+        for batch in valid_dataset:
+            character, mel, mel_input, pos_text, pos_mel, text_len = batch
+
+            t_ae_loss, s_ae_loss = autoencoder_step(model, batch)
+            losses['t_ae'].append(t_ae_loss.detach().cpu().item())
+            losses['s_ae'].append(s_ae_loss.detach().cpu().item())
+
+            asr_loss, tts_loss = supervised_step(model, batch)
+            losses['asr_'].append(asr_loss.detach().cpu().item())
+            losses['tts_'].append(tts_loss.detach().cpu().item())
+
+            t_cm_loss, s_cm_loss = crossmodel_step(model, batch)
+            losses['s_cm'].append(s_cm_loss.detach().cpu().item())
+            losses['t_cm'].append(t_cm_loss.detach().cpu().item())
+
+            text_pred = model.asr(None, mel.to(DEVICE), infer=True).squeeze()
+            len_mask_max, len_mask_idx = torch.max((text_pred == PAD_IDX), dim=1)
+            len_mask_idx[end_mask_max == 0] = text_pred.shape[1] - 1
+            print("Lengths", len_mask_idx.shape)
+            per += compute_per(character.to(DEVICE), text_pred, text_len.to(DEVICE), len_mask_idx)
+            n_iters += 1
+
+    model.train()
```

Comment:
I don't think the `model.train()` should be added here. This should be put in the `train()` function on line 155-158:
```python3
    ...
    for epoch in range(args.epochs):
        model.train()

        # Training model
        supervised_dataloader = DataLoader(supervised_train_dataset, batch_size=args.batch_size, shuffle=True, collate_fn=collate_fn_transformer, drop_last=True, num_workers=16)
        ...
```

Reply:
On another note, probably leave a comment here for later: `# TODO: probably evaluate speech somehow?`

Reply:
Good insight about the `model.train()` call, it definitely makes more sense in the "train" function.